File: Yandex/Practicum/Algorightms_and_Data_Structures/6_sprint/A._Построить_список_смежности/solution.py
import sys


def solution() -> None:
    """Time Complexity O((V + E) + ElogE)
    Memory Complexity O(V + E)"""
    v, e = map(int, sys.stdin.readline().rstrip().split())
    adj_list: dict[int, list[int]] = dict()
    for _ in range(e):
        source, destination = map(int, sys.stdin.readline().rstrip().split())
        if source not in adj_list:
            adj_list[source] = []
        adj_list[source].append(destination)
    for vertex in range(1, v + 1):
        edges: int = len(adj_list.get(vertex, []))
        if not edges:
            sys.stdout.write(str(edges) + '\n')
        else:
            adj_list[vertex].sort()
            sys.stdout.write(str(edges) + ' ' + ' '.join(str(neighbor) for neighbor in adj_list[vertex]) + '\n')


if __name__ == '__main__':
    solution()

File: Yandex/Practicum/Algorightms_and_Data_Structures/6_sprint/A._Построить_список_смежности/solution.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <algorithm>


void solution() {
    /*
    Time Complexity O((V + E) + ElogE)
    Memory Complexity O(V + E)
    */
    int v, e;
    std::cin >> v >> e;
    std::unordered_map<int, std::vector<int>> adjList;
    for (int i = 0; i < e; ++i) {
        int source, destination;
        std::cin >> source >> destination;
        adjList[source].push_back(destination);
    }
    for (int vertex = 1; vertex <= v; ++vertex) {
        int edges = adjList[vertex].size();
        if (!edges) std::cout << 0 << "\n";
        else {
            std::sort(adjList[vertex].begin(), adjList[vertex].end());
            std::cout << edges << " ";
            for (int idx = 0; idx < adjList[vertex].size(); ++idx) {
                if (idx) std::cout << " ";
                std::cout << adjList[vertex][idx];
            }
            std::cout << "\n";
        }
    }
}


int main() {
    solution();
}